### PR TITLE
DPDK flow dump

### PR DESCRIPTION
**Pull request by sjwahub**
Signed-off-by: Shujun Wang <shujun.wang@intel.com>

Flow dump automation test cases.

Testing log  [ptf_tests_run_flow3.log](https://github.com/ipdk-io/ptf-tests/files/10705685/ptf_tests_run_flow3.log)

### CODE REVIEW

File: common/utils/p4rtctl_utils.py [hunk 1]
```diff
@@ -387,3 +386,20 @@ def p4rt_ctl_reset_counter_entry(bridge, cnt_tbl_name, flow):
            ipv4_host_tbl_flow_counter_packets counter_id=303591076,index=1
     """
     p4rt_ctl.p4ctl_reset_counter_entry(bridge, cnt_tbl_name, flow)
+
+def p4rt_ctl_dump_entities(bridge):
+    try:
```
Comment (dkaranjai) [inline common/utils/p4rtctl_utils.py:395]: function definition missing
refer above functions

File: tests/DPDK_Flow_Dump_Help.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+        # List the p4rt_ctl Help menu 
+        for option in self.config_data["help_option"]:
+            helpcmd = f"p4rt-ctl {option}"
+            log.info(f'To verify command "{helpcmd}"')
+            result = test_utils.get_p4rt_ctl_help(option)
+            if result:
+                for each in self.config_data["arguments"]:
+                    if each in result:
+                        log.passed(
+                            f'PASS: the argument "{each}" is verifed in "{helpcmd}"'
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_Help.py]: you dont need write PASS: or FAIL: log.passed and log.failed does that

File: tests/DPDK_Flow_Dump_Help.py [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    def setUp(self):
+        BaseTest.setUp(self)
+        self.result = unittest.TestResult()
+
+        test_params = test_params_get()
+        config_json = test_params["config_json"]
+        self.config_data = get_config_dict(config_json)
+
+    def runTest(self):
+        # List the p4rt_ctl Help menu
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_Help.py:46]: add more information on what is config_data["help_option"] and ["arguments"] just an example

File: tests/DPDK_Flow_Dump_with_LNT.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        self.config_data = get_config_dict(
+            config_json,
+            pci_bdf=test_params["lnt_pci_bdf"],
+        )
+        
+        self.gnmictl_params = get_gnmi_params_simple(self.config_data)
+        self.gnmictl_hotplug_params = get_gnmi_params_hotplug(self.config_data)
+        self.tap_port_list = gnmi_ctl_utils.get_tap_port_list(self.config_data)
+        self.link_port_list = gnmi_ctl_utils.get_link_port_list(self.config_data)
+        log.info(self.gnmictl_hotplug_params)
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_with_LNT.py]: this can be removed

File: tests/DPDK_Flow_Dump_with_LNT.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+        self.interface_ip_list = get_interface_ipv4_dict(self.config_data)
+        self.conn_obj_list = []
+
+    def runTest(self):
+        log.info(f"Begin to configure OVVS and VM on local host")
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_with_LNT.py]: f not needed

File: tests/DPDK_Flow_Dump_with_LNT.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+            self.config_data["vxlan"]["tep_ip"][1].split("/")[0],
+            self.config_data["vxlan"]["dst_port"][0],
+        ):
+
+            self.result.addFailure(self, sys.exc_info())
+            self.fail(
+                f"Failed to add vxlan {self.config_data['vxlan']['vxlan_name'][0]} to bridge {self.config_data['bridge']}"
+            )
+
+        for i in range(len(self.conn_obj_list)):
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_with_LNT.py]: where are you populating conn_obj_list, its set to [] in setup(), please describe the purpose of this variable, maybe in setup()

File: tests/DPDK_Flow_Dump_with_LNT.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        log.info("Get flow dump table entry")
+        dump_table = p4rt_ctl.p4rt_ctl_dump_entities(self.config_data['switch'])
+        # Verify each entry
+        log.info("Verify each entry")
+        if len(dump_table[1:-1]) != len(self.config_data['flow_dump_table']):
+            log.failed(f"The number of {len(dump_table[1:-1])} is different than \
+                                 definition {len(self.config_data['flow_dump_table'])}")
+            log.info(f"The table is {dump_table[1:-1]}")
+            self.result.addFailure(self, sys.exc_info())
+            self.fail(f"Failed to verify flow dump entries\"")
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_with_LNT.py]: please scan and remove unnecessary 'f'

File: tests/DPDK_Flow_Dump_with_PNA_indirect_counter.py [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+                    self.result.addFailure(self, sys.exc_info())
+                    self.fail(f"Failed to add table entry {match_action}")
+        
+        # Get dump entries table
+        log.info("Get flow dump table entry")
+        dump_table = p4rt_ctl.p4rt_ctl_dump_entities(self.config_data['switch'])
+        # Verify each entry
+        log.info("Verify each table entry")
+        if len(dump_table[1:-1]) != len(self.config_data["flow_dump_table"]):
+            log.failed(f"The table {dump_table[1:-1]} has different entry than definition")
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_with_PNA_indirect_counter.py]: self.fail will also print failure message. see if this is needed

File: tests/DPDK_Flow_Dump_with_PNA_indirect_counter.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+            self.result.addFailure(self, sys.exc_info())
+            self.fail(f"Failed to verify flow dump entries\"")
+        else:
+            for each in dump_table[1:-1]:
+                entry = each.strip()
+                if entry in self.config_data["flow_dump_table"]:
+                    log.passed(f"The entry \"{entry}\" in flow dump is verified")
+                else:
+                    log.failed(f"The entry \"{entry}\" in flow dump is not verified")
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_with_PNA_indirect_counter.py]: same as above

File: tests/DPDK_Flow_Dump_with_action_selector.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+            log.failed(f"The table {dump_table[1:-1]} has different entry than definition")
+            self.result.addFailure(self, sys.exc_info())
+            self.fail(f"Failed to verify flow dump entries\"")
+        else:
+            for each in dump_table[1:-1]:
+                entry = each.strip()
+                if entry in self.config_data["flow_dump_table"]:
+                    log.passed(f"The entry \"{entry}\" in flow dump is verified")
+                else:
+                    log.failed(f"The entry \"{entry}\" in flow dump is not verified")
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_with_action_selector.py]: same as above

File: tests/DPDK_Flow_Dump_with_action_selector.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+                table["switch"], table["name"], match_action
+            ):
+                self.result.addFailure(self, sys.exc_info())
+                self.fail(f"Failed to add table entry {match_action}")
+      
+        # Get dump entries table
+        dump_table = p4rt_ctl.p4rt_ctl_dump_entities(self.config_data['switch'])
+        # Verify each entry
+        if len(dump_table[1:-1]) != len(self.config_data["flow_dump_table"]):
+            log.failed(f"The table {dump_table[1:-1]} has different entry than definition")
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_with_action_selector.py]: same as above

File: tests/DPDK_Flow_Dump_with_del_add.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+            log.failed(f"The table {dump_table[1:-1]} has different entry than definition")
+            self.result.addFailure(self, sys.exc_info())
+            self.fail(f"Failed to verify flow dump entries\"")
+        else:
+            for each in dump_table[1:-1]:
+                entry = each.strip()
+                if entry in self.config_data["flow_dump_table"]:
+                    log.passed(f"The entry \"{entry}\" in flow dump is verified")
+                else:
+                    log.failed(f"The entry \"{entry}\" in flow dump is not verified")
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_with_del_add.py]: same as above

File: tests/DPDK_Flow_Dump_with_del_add.py [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+                    table["switch"], table["name"], match_action
+                ):
+                    self.result.addFailure(self, sys.exc_info())
+                    self.fail(f"Failed to add table entry {match_action}")
+        
+        log.info("Verify flow dump table after adding back rules")
+        # Verify each entry        
+        dump_table = p4rt_ctl.p4rt_ctl_dump_entities(self.config_data['switch'])
+        if len(dump_table[1:-1]) != len(self.config_data["flow_dump_table"]):
+            log.failed(f"The table {dump_table[1:-1]} has different entry than definition")
```
Comment (dkaranjai) [inline tests/DPDK_Flow_Dump_with_del_add.py]: same as above

**Review by dkaranjai — CHANGES_REQUESTED** [review]

File: tests/DPDK_Flow_Dump_with_LNT.py (hunk 4 above)
Comment (sjwahub) [reply tests/DPDK_Flow_Dump_with_LNT.py]: Hotplug need to be removed. but link_port and tap_list needed for vxlan configuration

File: tests/DPDK_Flow_Dump_with_PNA_indirect_counter.py (hunk 9 above)
Comment (sjwahub) [reply tests/DPDK_Flow_Dump_with_PNA_indirect_counter.py]: The statements have a variables {entry}. The f is needed

File: tests/DPDK_Flow_Dump_with_action_selector.py (hunk 10 above)
Comment (sjwahub) [reply tests/DPDK_Flow_Dump_with_action_selector.py]: The statements have a variables {entry}. The f is needed

File: tests/DPDK_Flow_Dump_with_action_selector.py (hunk 11 above)
Comment (sjwahub) [reply tests/DPDK_Flow_Dump_with_action_selector.py]: The statements have a variables {entry}. The f is needed

File: tests/DPDK_Flow_Dump_with_del_add.py (hunk 13 above)
Comment (sjwahub) [reply tests/DPDK_Flow_Dump_with_del_add.py]: The statements have a variables {entry}. The f is needed

File: tests/DPDK_Flow_Dump_Help.py (hunk 3 above)
Comment (sjwahub) [reply tests/DPDK_Flow_Dump_Help.py:46]: done

**Comment by sjwahub** [comment]
Running log after changing code based on comments  [ptf_tests_run_flow4.log](https://github.com/ipdk-io/ptf-tests/files/10717750/ptf_tests_run_flow4.log)

**Review by dkaranjai — APPROVED** [review]
LGTM

**Pull request merged**